update_realm: Allow demo orgs to be converted to regular orgs.

eeshangarg opened the pull request:
This commit adds support to the `PATCH /realm` endpoint for
converting a demo organization to a regular organization.

This is a part of #19523.

@timabbott FYI :)

eeshangarg commented:
@timabbott: I just pushed another commit that adds a banner to the `Organization profile` settings section. Note that the conversion `dialog_widget` hasn't been added yet, that will be added in a separate commit. This area of the code is very new to me so I wanted to get feedback on the visuals first. This is what the banner looks like:

![demo_org_warning](https://user-images.githubusercontent.com/7251823/133653750-2c95b16c-4300-4ae1-978e-24a6fd5917ec.PNG)

This is now waiting on a review from you.

timabbott commented on zerver/views/realm.py:
```diff
@@ -158,6 +167,19 @@ def update_realm(
+            raise JsonableError(_("Organization subdomain is required."))
+
+        if string_id is not None and realm.demo_organization_scheduled_deletion_date is not None:
+            do_convert_demo_realm_to_regular_realm(realm, string_id)
```
This needs to call `check_subdomain_available` -- otherwise you can override another organization's subdomain, or a restricted one.

From an API perspective, I think we could just have the a `string_id` parameter -- it's implied that if you're a demo organization and want to change your subdomain, that you want to convert to a real organization, and lets us simplify the API.  

(Also, if we start allowing folks to change their subdomain, it'd be able to use the same interface).

eeshangarg replied:
Ooh that's a great idea!

timabbott commented on zerver/views/realm.py:
```diff
@@ -158,6 +167,19 @@ def update_realm(
+            raise OrganizationOwnerRequired()
+
+        if realm.demo_organization_scheduled_deletion_date is None:
+            raise JsonableError(_("Organization to be converted must be a demo organization."))
```
I think "Only demo organizations can use this endpoint" is probably simpler copy.

timabbott commented:
We probably want to write code to redirect the browser to the new URL after such a domain change.  Because the events system will stop working the moment that `realm.string_id` changes, I believe that means we should make the `success` handler do that reload.  That probably means we want the HTTP response to have a `realm_url` key in it with the new `realm.host`.

eeshangarg commented:
@timabbott I addressed your comments. The first two commits are ready for review, the third one needs a bit of work still. Thanks!

eeshangarg commented:
@timabbott All three commits are now ready for review.

timabbott commented:
I merged the first commit after these changes:

```
diff --git a/templates/zerver/api/changelog.md b/templates/zerver/api/changelog.md
index 5a406903c6..643f0cdec8 100644
--- a/templates/zerver/api/changelog.md
+++ b/templates/zerver/api/changelog.md
@@ -11,6 +11,12 @@ below features are supported.
 
 ## Changes in Zulip 5.0
 
+**Feature level 104**
+
+* [`PATCH /realm`]: Added `string_id` parameter for changing an
+  organization's subdomain. Currently, this is only allowed for
+  changing changing a demo organization to a normal one.
+
 **Feature level 103**
 
 * [`POST /register`](/api/register-queue): Added `create_web_public_stream_policy`
diff --git a/version.py b/version.py
index ac6b519911..5d2d4b2aef 100644
--- a/version.py
+++ b/version.py
@@ -33,7 +33,7 @@ DESKTOP_WARNING_VERSION = "5.4.3"
 # Changes should be accompanied by documentation explaining what the
 # new level means in templates/zerver/api/changelog.md, as well as
 # "**Changes**" entries in the endpoint's documentation in `zulip.yaml`.
-API_FEATURE_LEVEL = 103
+API_FEATURE_LEVEL = 104
 
 # Bump the minor PROVISION_VERSION to indicate that folks should provision
 # only when going from an old version of the code to a newer version. Bump
diff --git a/zerver/tests/test_realm.py b/zerver/tests/test_realm.py
index f32aa2e35e..c959c8bf62 100644
--- a/zerver/tests/test_realm.py
+++ b/zerver/tests/test_realm.py
@@ -171,45 +171,29 @@ class RealmTest(ZulipTestCase):
         realm = get_realm("zulip")
         self.assertNotEqual(realm.description, new_description)
 
-    def test_realm_convert_demo_org_not_a_realm_owner(self) -> None:
+    def test_realm_convert_demo_realm(self) -> None:
         data = dict(string_id="coolrealm")
 
         self.login("iago")
-
         result = self.client_patch("/json/realm", data)
         self.assert_json_error(result, "Must be an organization owner")
 
-    def test_realm_convert_regular_realm(self) -> None:
-        data = dict(string_id="coolrealm")
-
         self.login("desdemona")
-
         result = self.client_patch("/json/realm", data)
         self.assert_json_error(result, "Must be a demo organization.")
 
-    def test_realm_convert_demo_realm_subdomain_not_available(self) -> None:
-        data = dict(string_id="zulipdev")
-
+        data = dict(string_id="lear")
         self.login("desdemona")
         realm = get_realm("zulip")
         realm.demo_organization_scheduled_deletion_date = timezone_now() + datetime.timedelta(
             days=30
         )
         realm.save()
-
         result = self.client_patch("/json/realm", data)
         self.assert_json_error(result, "Subdomain unavailable. Please choose a different one.")
 
-    def test_realm_convert_demo_realm_successfully(self) -> None:
-        data = dict(is_demo_organization=orjson.dumps(False).decode(), string_id="coolrealm")
-
-        self.login("desdemona")
-        realm = get_realm("zulip")
-        realm.demo_organization_scheduled_deletion_date = timezone_now() + datetime.timedelta(
-            days=30
-        )
-        realm.save()
-
+        # Now try to change the string_id to something available.
+        data = dict(string_id="coolrealm")
         result = self.client_patch("/json/realm", data)
         self.assert_json_success(result)
         json = orjson.loads(result.content)
diff --git a/zerver/views/realm.py b/zerver/views/realm.py
index 271b9c105e..c160a4ac77 100644
--- a/zerver/views/realm.py
+++ b/zerver/views/realm.py
@@ -278,21 +278,20 @@ def update_realm(
         else:
             data["default_code_block_language"] = default_code_block_language
 
-    if string_id:
+    if string_id is not None:
         if not user_profile.is_realm_owner:
             raise OrganizationOwnerRequired()
 
         if realm.demo_organization_scheduled_deletion_date is None:
             raise JsonableError(_("Must be a demo organization."))
 
-        if string_id is not None and realm.demo_organization_scheduled_deletion_date is not None:
-            try:
-                check_subdomain(string_id)
-            except ValidationError as err:
-                raise JsonableError(str(err.message))
+        try:
+            check_subdomain(string_id)
+        except ValidationError as err:
+            raise JsonableError(str(err.message))
 
-            do_convert_demo_realm_to_regular_realm(realm, string_id)
-            data["realm_uri"] = realm.uri
+        do_convert_demo_realm_to_regular_realm(realm, string_id)
+        data["realm_uri"] = realm.uri
 
     return json_success(data)
 
```

* The testing cleanup is motivated by https://zulip.readthedocs.io/en/latest/testing/philosophy.html#share-test-setup-code.  I also noticed while fixing it that we were passing a parameter that no longer exists in the success case, which is the kind of detail that gets messed up when copy-pasting test setup code.
* We need to document all API changes in public APIs, and `POST /realm` is one of those, so I added that.
* I removed some unnecessary conditionals in the view code, which I think led to a latent bug that `string_id=""` would not work, and also just repeated content.
* (Not shown in diff above) I switched to using `do_change_realm_subdomain` rather than adding a new function that partially duplicated it but lacked `RealmAuditLog` update logic.

I then merged the first commit as 29b354346b30cf1492a7dfc8c2b0866e2f3cc65d.

timabbott commented:
That language sgtm.

eeshangarg commented:
@alya I spent some more time on this PR today. As per our discussion last week, the WIP commit I just pushed changed the CSS of the banner to act link a button, instead of just "Convert organization" being a button inside of the banner. Could you please review it? Here are the screenshots:

Without hovering:
![demo-org-warning](https://user-images.githubusercontent.com/7251823/140767458-12528727-b950-4cef-b0e7-14581729d046.png)

When hovering:
![demo-org-warning-hover](https://user-images.githubusercontent.com/7251823/140767530-7f403d84-4e89-444b-a60c-88ee0afee579.png)

A couple of more questions:
* I propose making the text inside the banner shorter, for brevity's sake. Brevity is important here so that the banner is as succinct as possible and the text doesn't end up being wrapped over multiple lines, which looks a little odd visually. Maybe "This demo organization will be deleted in 30 days. Click this banner to make organization permanent." Even this seems a bit too long.
* As for having a confirmation dialog once the user slicks "Make organization permanent", that would lead to two pop-ups on top of the settings pop-up, which seems like it might be a bit overwhelming visually? Should we maybe just have one dialog and mention the permanence of this decision in a warning banner inside the first dialog?

Let me know what you think! Thanks! :)

alya commented:
1) The whole banner should be a link, and the mouse pointer needs to be a hand when you hover over the banner. I don't see that happening, unless I have the wrong version of the PR or something?
2) Are the colors the same as on the yellow upgrade your org banner? The text looks darker to me than on that one, but it might be my imagination. Anyway, they should be the same.
3) The CTA shouldn't have link formatting since the whole banner is a link.
4) For the text, we should match the style to the upgrade banner CTA. How about this for the second sentence: "Convert to make permanent."?

alya commented:
> As for having a confirmation dialog once the user slicks "Make organization permanent", that would lead to two pop-ups on top of the settings pop-up, which seems like it might be a bit overwhelming visually? Should we maybe just have one dialog and mention the permanence of this decision in a warning banner inside the first dialog?

Hm, I'm not sure. It might be all right with the fading we do? We could try it.

alya commented:
Maybe "Convert to make it permanent." would be more clear than "Convert to make permanent." ?

alya commented:
Since we discussed making demo org creation one-click, I think we'll need a much more complete form. Otherwise, we aren't collecting the rest of the realm creation info, right?
<img width="398" alt="Screen Shot 2021-11-12 at 2 58 28 PM" src="https://user-images.githubusercontent.com/2090066/141590671-4f702790-5539-49bd-b457-c7d4b026e893.png">

eeshangarg commented:
@alya: As per our discussion about redirecting to `/new` instead of having a dialog flow, I just updated the sole commit in this PR to only add a banner that has links and is no longer "a button" itself. Here is what it looks like:
![demo_org_warning](https://user-images.githubusercontent.com/7251823/142214363-60c035ea-dbfb-46a9-a6c7-6a445e213d0e.png)

Let me know what you think! Thanks! :)

alya commented:
OK, here's a proposal for the conversion flow.

First screen, which is shown only if the user doesn't have an account yet:

<img width="876" alt="convert_org1" src="https://user-images.githubusercontent.com/2090066/143093529-04a2bc2a-0a85-4caa-a5cd-21be1ca20892.png">
.

alya commented:
Second screen:
<img width="1269" alt="convert_org2" src="https://user-images.githubusercontent.com/2090066/143093625-475db92b-ec48-47cd-bbdb-48462a002596.png">

eeshangarg commented:
@alya Thanks for the feedback, that is super helpful! I did push some of my code that takes a stab at implementing your suggestion for the first page for email confirmations, along with the backend changes. 

As for the import line, I am not sure if it belongs on the demo org conversion page, wouldn't it be better to keep it simpler? It seems like the person who created a demo org would be highly unlikely to also want to import stuff from another tool at the same time that they convert the org. That seems like a great feature to add eventually, but if we don't already have the ability to import data and convert the org simultaneously via the web app, it shouldn't be advertised on the conversion page. What do you think? 

Here is a screenshot of what the email submission form looks like now:

![org_conversion_email_confirmation](https://user-images.githubusercontent.com/7251823/143724743-69127e66-00a8-4e83-b32f-127cfc99976c.png)

@timabbott Could you please take a look at some of the backend changes and tell me what you think about me passing around the user profile ID in URLs to keep track of which demo realm is being converted? This approach seems less than elegant to say the least, and I was wondering if you had any suggestions for accomplishing this in a more elegant way?

alya commented:
Err, the "will this work? ..." part was a question from me about the import line, not text to put on the page. :)

alya commented:
Actually, looking at the conversion instructions, we should keep the same line as we have on the new org page. The instructions all say to email support for Zulip Cloud orgs, so it doesn't matter whether you already have a demo org or not.

I want to avoid the situation where people realize too late they wanted to do an import, and end up with useful data both in the other chat tool and in Zulip that is impossible to merge.

eeshangarg commented:
@timabbott I just resolved conflicts. Also, we could potentially merge the first commit in this PR, but the second one needs more work. Thanks! :)

eeshangarg commented:
@alya: Just pushed a commit that disables billing pages for demo orgs. Screenshot:
![demo-org-billing-error](https://user-images.githubusercontent.com/7251823/147507164-1e9e94ed-f0e8-487f-96b7-be1065e7469b.png)

alya commented:
> @alya: Just pushed a commit that disables billing pages for demo orgs. Screenshot: ![demo-org-billing-error](https://user-images.githubusercontent.com/7251823/147507164-1e9e94ed-f0e8-487f-96b7-be1065e7469b.png)

Hm, that looks pretty weird by itself. One option might be to keep the pages as they normally are, but:

1) Add that banner just below the "🚀 Upgrade to Zulip Standard" heading.
2) Disable all the action buttons ("Add card", "Buy standard" and "Submit").

Thoughts?

alya commented:
Also, while we're here, could you submit a quick PR to make the font on all three of those buttons ("Add card", "Buy standard" and "Submit") the same size? It can be the same size as the "We'll automatically charge you for additional licenses as users are added, and remove licenses not in use at the end of each billing period," text.

eeshangarg commented:
@alya I made the buttons on billing the same size and style, as you requested. Here is the latest screenshot:


![Screenshot 2022-01-14 191405](https://user-images.githubusercontent.com/7251823/149600542-7301bad7-8e9d-4fc9-ab22-f0734d5277a7.png)

eeshangarg commented:
@alya I just addressed your feedback for disabling billing for demo orgs. Here is an updated screenshot:
![demo-orgs-upgrade-page](https://user-images.githubusercontent.com/7251823/149830332-515e23b5-90d3-402a-aeb5-9e4388a268eb.png)

alya commented:
Cool, thanks!

What's the hover behavior for the "Add card" button? We should probably have a disabled cursor with a tooltip that says something like, "Convert demo organization before upgrading."

eeshangarg commented:
@alya Just addressed your feedback and added a disabled cursor for the buttons and a tooltip:
![tooltip](https://user-images.githubusercontent.com/7251823/150031689-05b43e74-b78d-4d1e-ba5c-f386dffc4a0f.png)

@timabbott The first three commits of this PR are ready for a review! Thanks!

eeshangarg commented:
Ah the build is failing, gonna take a look at this tomorrow morning!

eeshangarg commented:
@timabbott Fixed the build issue, so the first three commits should be good to review and possibly merge! Thanks! :)

timabbott commented:
I merged 3b9455c9cebf91df69a7b325eb64222d394d52e1 (aka the button size commit).

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5407,6 +5407,11 @@ def do_create_realm(
         kwargs["enable_spectator_access"] = enable_spectator_access
 
+    # Welcome emails are disabled for demo orgs since their owners
+    # may not have a valid email set.
```
s/may/will/, right?  Also, I'd call it "demo organizations" -- part of our "no abbreviations" policy.

eeshangarg replied:
Done!

timabbott commented on zerver/views/registration.py:
```diff
@@ -155,6 +160,7 @@ def accounts_register(
     source_realm_id: Optional[int] = REQ(
         default=None, converter=to_converted_or_fallback(to_non_negative_int, None)
     ),
+    is_demo_organization: Optional[bool] = REQ(converter=to_bool, default=False),
```
I think you want to use `validator=check_bool`, not `converter`; at least that's what essentially all of our other endpoints do.  And then we don't need to add a new duplicate validation function.

eeshangarg replied:
We actually can't use `validator=check_bool` here. I am pretty sure `check_bool` is only used as `json_validator=check_bool`. The request for creating a realm is not a JSON request. If you look at [the POST request we make for realm creation](https://github.com/zulip/zulip/blob/main/zerver/views/development/registration.py#L103), the booleans `terms` and `is_demo_organization` are passed as strings `"true"`. When you change the above code to use `validator=check_bool`, `mypy` complains with the following:

```
mypy      | zerver/views/registration.py:164: error: No overload variant of "REQ" matches argument types "Callable[[str, object], bool]", "bool"  [call-overload]
mypy      | zerver/views/registration.py:164: note: Possible overload variants:
mypy      | zerver/views/registration.py:164: note:     def [ResultT] REQ(whence: Optional[str] = ..., *, converter: Callable[[str], ResultT], default: ResultT = ..., intentionally_undocumented: bool = ..., documentation_pending: bool = ..., aliases: Sequence[str] = ..., path_only: bool = ...) -> ResultT
mypy      | zerver/views/registration.py:164: note:     def [ResultT] REQ(whence: Optional[str] = ..., *, default: ResultT = ..., json_validator: Callable[[str, object], ResultT], intentionally_undocumented: bool = ..., documentation_pending: bool = ..., aliases: Sequence[str] = ..., path_only: bool = ...) -> ResultT
mypy      | zerver/views/registration.py:164: note:     def REQ(whence: Optional[str] = ..., *, default: str = ..., str_validator: Optional[Callable[[str, object], str]] = ..., intentionally_undocumented: bool = ..., documentation_pending: bool = ..., aliases: Sequence[str] = ..., path_only: bool = ...) -> str
mypy      | zerver/views/registration.py:164: note:     def REQ(whence: Optional[str] = ..., *, default: None, str_validator: Optional[Callable[[str, object], str]] = ..., intentionally_undocumented: bool = ..., documentation_pending: bool = ..., aliases: Sequence[str] = ..., path_only: bool = ...) -> Optional[str]
mypy      | zerver/views/registration.py:164: note:     def [ResultT] REQ(whence: Optional[str] = ..., *, default: ResultT = ..., argument_type: Literal['body'], intentionally_undocumented: bool = ..., documentation_pending: bool = ..., aliases: Sequence[str] = ..., path_only: bool = ...) -> ResultT
mypy      |
mypy      | See https://zulip.readthedocs.io/en/latest/testing/mypy.html for debugging tips.
```

So the only way to convert these strings into a `bool` was to add a dedicated converter here. It could be that I am missing something here, let me know what you think! Thanks! :)

timabbott replied:
Hmm.  I don't we want to keep the `validator.py` library as a pure library for things that match the `json_validator` system.  @andersk do you have a recommendation for how we should validate a boolean for this function?

timabbott commented on zerver/views/registration.py:
```diff
@@ -172,7 +178,13 @@ def accounts_register(
+        # upon demo realm creation, so we can skip email validation
+        # for demo org owners.
+        if realm_creation and is_demo_organization:
+            pass
```
I think if we're skipping validation, we should do the if statement outside the try/except block, right?  And potentially have explicit code that discards `prereg_user.email` if `is_demo_organization && realm_creation`, to ensure we never write an unvalidated email address to the database?

timabbott commented on templates/corporate/upgrade.html:
```diff
@@ -24,6 +24,16 @@ <h1>🚀 {% trans %}Upgrade to {{ plan }}{% endtrans %}</h1>
+                    <a href="/help/demo-organizations#converting-a-demo-organization-to-a-regular-organization">
+                        convert your demo organization
+                    </a>
+                    before upgrading.
```
@alya can you propose a revision to this string?

timabbott replied:
I think this billing/upgrade page commit is probably pretty close to mergeable, not having reviewed the CSS at all. I still think it's worth make it into its own independent PR (perhaps split into two commits, the first one adding Tippy configuration for the portico, and the second being the rest of this), since it's unrelated to the other changes in this PR, and it'll have some back-and-forth about wording and the like before it can be merged.

@laurynmm FYI.

timabbott commented on templates/corporate/upgrade.html:
```diff
@@ -247,7 +264,11 @@ <h4>Describe your organization briefly</h4>
                                 <br />
                                 <p id="sponsorship-discount-details"></p>
-                                <button type="submit" id="sponsorship-button" class="stripe-button-el invoice-button">Submit</button>
+                                <div class="upgrade-button-container" {% if is_demo_org %}data-tippy-content="{% trans %}Convert demo organization before upgrading.{% endtrans %}"{% endif %}>
```
Why do we need this new container div?

eeshangarg replied:
Because disabled elements do not fire any events and thus `tippyjs` will not work with disabled elements. Will add a comment explaining that.

timabbott replied:
@amanagr @sahil839 -- are you aware of a cleaner way to handle giving Tippy tooltips to disabled elements?  I'm OK if this is necessary, but I feel like it's a pattern we have in settings and want to make sure we're not missing something.

amanagr replied:
I am not sure if there are much cleaner ways.

eeshangarg commented:
@timabbott Just addressed your feedback, the first two commits are ready for another review. Thanks! :)

eeshangarg commented:
@timabbott The first two commits are ready for merging, the third commit needs more work. Also, the checklist on #19523 has been updated to reflect the latest plan for demo orgs. Thanks!

timabbott commented on static/js/billing/upgrade.js:
```diff
@@ -71,6 +72,18 @@ export const initialize = () => {
+            animation: false,
+            touch: ["hold", 750],
+            placement: "bottom",
+        });
```
Couldn't we just make this library initializing unconditional?  

I guess there's a question of whether we want to be doing this in something like `portico.js` rather than a billing-specific page; @amanagr thoughts?

amanagr replied:
True, let's just initialize this in portico.

timabbott replied:
OK, so let's split this out as an independent prep commit that adds Tippy support for portico pages; perhaps @amanagr and @laurynmm can coordinate on having that happen.

timabbott commented on zerver/actions/create_realm.py:
```diff
@@ -167,6 +167,10 @@ def do_create_realm(
+    if is_demo_organization:
+        # Welcome emails are disabled for demo organizations, since
+        # their owners will not have a valid email set.
+        kwargs["send_welcome_emails"] = False
```
This is the wrong way to do this; we likely don't want to change the default for sending welcome emails in the organization in general -- just block them for the new owner of this demo organization.

I think probably we should replace this with a change inside `enqueue_welcome_emails` of the form:
```
if user_profile.delivery_email == "":
    # Do not attempt to enqueue welcome emails for users without an email address.
    # The assertions here are to help document the only circumstance under which
    # this condition should be possible.
    assert user_profile.realm.is_demo_organization and realm_creation
    return
```

@laurynmm probably this commit can just be replaced with an independent PR with that revised plan.

timabbott replied:
See also https://github.com/zulip/zulip/pull/19741/commits/f662b0834cb157e4ba161534ee88290a28fa55c3#r932756780 - I think that actually we're going to end up skipping every part of this commit, and just doing the thing described in the last comment.

timabbott commented on zerver/lib/validator.py:
```diff
@@ -588,6 +588,24 @@ def to_timezone_or_empty(var_name: str, s: str) -> str:
+    else:
+        value = bool(s)
+
+    return value
```
Ideally, we want to avoid adding a new function like this. https://github.com/zulip/zulip/pull/19741/commits/f662b0834cb157e4ba161534ee88290a28fa55c3#r788252405 has some more discussion. I don't have a specific solution in mind, but I'm not sure we actually need a boolean parameter in this endpoint at all.

timabbott commented on zerver/views/registration.py:
```diff
@@ -170,10 +176,17 @@ def accounts_register(
+        try:
+            validators.validate_email(email)
+        except ValidationError:
+            return render(request, "zerver/invalid_email.html", context={"invalid_email": True})
```
I think we don't want this hunk yet, since our plan is to avoid making it possible to create demo organizations except via the development environment button, and this would do so.

timabbott commented on static/styles/portico/portico_signin.css:
```diff
@@ -401,6 +401,33 @@ html {
         font-weight: 400;
     }
 
+    .demo-org-warning {
```
Just a note that we probably want to refer to these using more consistent terminology, in the variables and CSS; `demo-realm` or `demo-organization` seems clearly better than `demo-org`.

timabbott commented:
@alya @laurynmm I've commented on everything I note in the first two commits, which I think should each be turned into their own PR, since they can be merged independently of each other.

The last commit can be its own independent PR; it looks like it's not too distant from the final shape that we want, but I think probably the first N rounds of iteration for it will be with @alya to get the language/flow right. 

I think that can be worked on mostly independently from the work on having an "add your email address" barrier to conversion -- we can test the "add your email address" requirement with the "invite user" flow, and then having it block this flow as well would be just be calling the function for that purpose in a second place.

timabbott commented:
Closing in favor of https://github.com/zulip/zulip/pull/22729.
